tests/Propulsion.Kafka0.Integration/ParallelIntegration.fs

```diff
@@ -0,0 +1,327 @@
+﻿namespace Propulsion.Kafka.Integration.Parallel
+
+open Jet.ConfluentKafka.FSharp
+open Newtonsoft.Json
+open Propulsion.Kafka
+open Serilog
+open Swensen.Unquote
+open System
+open System.Collections.Concurrent
+open System.ComponentModel
+open System.Threading
+open System.Threading.Tasks
+open Xunit
+
+[<AutoOpen>]
+[<EditorBrowsable(EditorBrowsableState.Never)>]
+module Helpers =
+    open Confluent.Kafka
+
+    // Derived from https://github.com/damianh/CapturingLogOutputWithXunit2AndParallelTests
+    // NB VS does not surface these atm, but other test runners / test reports do
+    type TestOutputAdapter(testOutput : Xunit.Abstractions.ITestOutputHelper) =
+        let formatter = Serilog.Formatting.Display.MessageTemplateTextFormatter("{Timestamp:yyyy-MM-dd HH:mm:ss.fff zzz} [{Level}] {Message}{NewLine}{Exception}", null);
+        let writeSerilogEvent logEvent =
+            use writer = new System.IO.StringWriter()
+            formatter.Format(logEvent, writer);
+            writer |> string |> testOutput.WriteLine
+            writer |> string |> System.Diagnostics.Debug.Write
+        interface Serilog.Core.ILogEventSink with member __.Emit logEvent = writeSerilogEvent logEvent
+
+    let createLogger sink =
+        LoggerConfiguration()
+            .Destructure.FSharpTypes()
+            .WriteTo.Sink(sink)
+            .WriteTo.Seq("http://localhost:5341")
+            .CreateLogger()
+
+    let getTestBroker() = 
+        match Environment.GetEnvironmentVariable "TEST_KAFKA_BROKER" with
+        | x when String.IsNullOrEmpty x -> invalidOp "missing environment variable 'TEST_KAFKA_BROKER'"
+        | x -> Uri x
+
+    let newId () = let g = System.Guid.NewGuid() in g.ToString("N")
+
+    type Async with
+        static member ParallelThrottled degreeOfParallelism jobs = async {
+            let s = new SemaphoreSlim(degreeOfParallelism)
+            return!
+                jobs
+                |> Seq.map (fun j -> async {
+                    let! ct = Async.CancellationToken
+                    do! s.WaitAsync ct |> Async.AwaitTask
+                    try return! j
+                    finally s.Release() |> ignore })
+                |> Async.Parallel
+        }
+
+    type ConsumerPipeline with
+        member c.StopAfter(delay : TimeSpan) =
+            Task.Delay(delay).ContinueWith(fun (_:Task) -> c.Stop()) |> ignore
+
+    type TestMessage = { producerId : int ; messageId : int }
+    [<NoComparison; NoEquality>]
+    type ConsumedTestMessage = { consumerId : int ; raw : ConsumeResult<string,string> ; payload : TestMessage }
+    type ConsumerCallback = ConsumerPipeline -> ConsumedTestMessage -> Async<unit>
+
+    let runProducers log (broker : Uri) (topic : string) (numProducers : int) (messagesPerProducer : int) = async {
+        let runProducer (producerId : int) = async {
+            let cfg = KafkaProducerConfig.Create("panther", broker, Acks.Leader)
```

Comment:
should this be a variable and not "panther" or is the test topic panther?

Reply:
this becomes the `client.id` and is also used in some ephemeral topic names in the test context (normally with a guid or id to guearantee uniqueness) - it could also be `"leopard"` or `"foo"` ;)